Update Container Wait

jlhawn opened the pull request:
**Summary**

The goal for this PR is to be able to use the `POST /containers/{name}/wait` API in order to no longer rely on the Events API when doing `docker run ...` and `docker start` in attached mode. Here is the set of behavioral changes in the API and CLI:

- The wait API now has on optional query parameter, `condition`, which specifies the state of the container which the client would like to wait for:

    - **`not-running`**

        This is the default value which maintains the existing behavior of the Wait API. It blocks if the container is in a state which is considered to be *running*, i.e., any of *running*, *paused*, or *restarting*. If it is already not in any of those states, then it would not block at all and return immediately with the current exit code.

    - **`next-exit`**

        This condition allows the client to wait for the container's state to *change* to a not-running state or to be removed. This is different from the *not-running* option because it will block if the container is in the *created*, *exited*, *dead*, or *removing* state, blocking through a transition to one of the running states and back to one of *exited* states or until the container is removed, whichever comes first. It is similar to waiting for a *die* or *destroy* event. This option is very useful to use before issuing a container start API request.

    - **`removed`**

        This condition allows the client to wait for the container to be removed. It is similar to waiting for a *destroy* event.

- The Wait API will now return HTTP response headers immediately. This serves an an acknowledgement that a wait has been registered. The JSON response body will not be returned until the wait condition is met.

- The Wait API's request context is extended to wait for a `http.CloseNotifier` event in the case that the client closes the connection while the server is blocked on the wait condition. This prevents leaked goroutines but should only be necessary until we switch to using Go 1.8's request context which cancels automatically if the client disconnects.

- ~~To handle a manual detach (when the user enters the detach keys sequence when attached in TTY mode), the CLI now detects this key sequence, stops sending stdin, and causes the CLI to exit immediately. This is done as an alternative to waiting for a *detach* event.~~ The rug has been pulled out from under this bullet point. This will have to be done in the github.com/docker/cli repo along with the other CLI changes.

Follow-up to #31794

Fixes #32242

Commit:
Update Container Wait Backend

This patch consolidates the two WaitStop and WaitWithContext methods
on the container.State type. Now there is a single method, Wait, which
takes a context and a bool specifying whether to wait for not just a
container exit but also removal.

The behavior has been changed slightly so that a wait call during a
Created state will not return immediately but instead wait for the
container to be started and then exited.

The interface has been changed to no longer block, but instead returns
a channel on which the caller can receive a *StateStatus value which
indicates the ExitCode or an error if there was one (like a context
timeout or state transition error).

These changes have been propagated through the rest of the deamon to
preserve all other existing behavior.

Docker-DCO-1.1-Signed-off-by: Josh Hawn <josh.hawn@docker.com> (github: jlhawn)

tonistiigi commented:
@WeiZhang555 @cpuguy83 any concerns for the approach?

WeiZhang555 commented:
The design looks good, I like it :-) 
https://github.com/docker/docker/pull/31744 seems need to reimplement based on this, it could be simplified a lot with this PR.

cpuguy83 commented:
@WeiZhang555 Indeed, however we still need to support old daemons.

WeiZhang555 commented:
@cpuguy83 Yep, you're right, then we have to keep both `waitExitOrRemoved` and this, backward compatibility makes this so complicated :cry:

Commit:
Update ContainerWait API

This patch adds the untilRemoved option to the ContainerWait API which
allows the client to wait until the container is not only exited but
also removed.

This patch also adds some more CLI integration tests for waiting for a
created container and waiting with the new --until-removed flag.

Docker-DCO-1.1-Signed-off-by: Josh Hawn <josh.hawn@docker.com> (github: jlhawn)

Handle detach sequence in CLI

Docker-DCO-1.1-Signed-off-by: Josh Hawn <josh.hawn@docker.com> (github: jlhawn)

Update Container Wait Conditions

Docker-DCO-1.1-Signed-off-by: Josh Hawn <josh.hawn@docker.com> (github: jlhawn)

Apply container wait changes to API 1.30

The set of changes to the containerWait API missed the cut for the
Docker 17.05 release (API version 1.29). This patch bumps the version
checks to use 1.30 instead.

This patch also makes a minor update to a testfile which was added to
the builder/dockerfile package.

Docker-DCO-1.1-Signed-off-by: Josh Hawn <josh.hawn@docker.com> (github: jlhawn)

Remove wait changes from CLI

Docker-DCO-1.1-Signed-off-by: Josh Hawn <josh.hawn@docker.com> (github: jlhawn)

Address minor nits on wait changes

- Changed the name of the tty Proxy wrapper to `escapeProxy`
- Removed the unnecessary Error() method on container.State
- Fixes a typo in comment (repeated word)

Docker-DCO-1.1-Signed-off-by: Josh Hawn <josh.hawn@docker.com> (github: jlhawn)

Use router.WithCancel in the containerWait handler

This handler previously added this functionality manually but now uses
the existing wrapper which does it for us.

Docker-DCO-1.1-Signed-off-by: Josh Hawn <josh.hawn@docker.com> (github: jlhawn)

Add WaitCondition constants to api/types/container

Docker-DCO-1.1-Signed-off-by: Josh Hawn <josh.hawn@docker.com> (github: jlhawn)

Address more ContainerWait review comments

- Update ContainerWait backend interface to not return pointer values
  for container.StateStatus type.
- Updated container state's Wait() method comments to clarify that a
  context MUST be used for cancelling the request, setting timeouts,
  and to avoid goroutine leaks.
- Removed unnecessary buffering when making channels in the client's
  ContainerWait methods.
- Renamed result and error channels in client's ContainerWait methods
  to clarify that only a single result or error value would be sent
  on the channel.

Docker-DCO-1.1-Signed-off-by: Josh Hawn <josh.hawn@docker.com> (github: jlhawn)

Move container.WaitCondition type to separate file

... to avoid conflict with swagger-generated code for API response

Docker-DCO-1.1-Signed-off-by: Josh Hawn <josh.hawn@docker.com> (github: jlhawn)

Address more ContainerWait review comments

Docker-DCO-1.1-Signed-off-by: Josh Hawn <josh.hawn@docker.com> (github: jlhawn)

jlhawn commented:
Well, at least `janky` seems happy with this first commit. I'll have to look into the failures with `windowsRS1` and `z` tests later. I'll push the second commit now.

jlhawn commented:
I've removed the *[WIP]* label from this PR and make the necessary changes to the API client and docs. Please take a look. I've also the `waitExitOrRemoved()` function in a way that maintains compatibility with older versions.

While all of the integration tests pass locally for me, I have seen some flakiness in a few tests that Jenkins runs. They mostly get this error:

```
Cannot kill container {ContainerID}: Container {ContinerID} is not running
```

But I can't tell if it's related to my changes or not since I tried not to modify container state/lifecycle in any way.

jlhawn commented on cli/command/container/utils.go:
```diff
@@ -13,12 +13,35 @@ import (
+			statusC <- int(result.StatusCode)
+		case err := <-errC:
+			logrus.Errorf("error waiting for container: %v", err)
+		}
```
Oops, I forgot to send do `statusC <- 125` here.

jlhawn commented:
I've hit an issue on something that I forgot to consider: `docker run -it ...` where the user enters the `DetachKeys`. In this case, the `ContainerAttach` API call returns but we still block on waiting for the container to exit, which it never will. This is why the `DockerSuite.TestAttachAfterDetach` integration test is failing everywhere.

The current implementation of `waitExitOrRemoved()` handles this by waiting for a `detach` event to unblock and return a `0` exit code. That is also not perfect because you have no way of knowing which CLI session the detach came from. I also want to completely avoid using the events API. I could overload the wait API with another option to wait for a detach, but you still don't know which attach request it was associated with.

One solution that I can think of would be to do an inspect of the container after the attach returns. If the container is still running then we can assume that the attach connection ended because the user entered the detach control keys. If the container has exited or is not found then we can assume that the attach connection ended because the container exited and/or was removed. The only problem with this is that there is a very small race window after the container exits between when the container's IO streams are closed (which causes the attach request to end) and the container's state is set to stopped/restarting. If the inspect request comes in after the attach connection closes but somehow before the container's state is marked as stopped then it would be interpreted as a manual detach since it would look like the container is still running.

Another clever solution to detecting the detach control keys is to actually do that detection client-side, passing the stream through the logic that tries to detect the detach control keys in the same way that the daemon-side does. Then we'll know for sure that the client initiated the detach.

jlhawn commented:
Windows tests look good. There's just a race condition in one of the DockerSuite integration tests. I've opened an issue for it. #32306

jlhawn commented:
The powerpc also looks good except for the same flaky test. It also failed on another flaky test: `TestSwarmIngressNetwork`.

jlhawn commented:
The `z` tests had the same single flaky test failure as on Windows: `TestRmRestartingContainer`.

jlhawn commented:
I've decided to carry a fix for `TestRmRestartingContainer` on top of this PR to try to get all of the tests to pass. I've also got another PR open with just that commit: https://github.com/docker/docker/pull/32307

jlhawn commented:
Aw Yiss. ✅✅✅✅✅✅

jlhawn commented:
I've edited the initial PR comment above to summarize all of the changes.

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "update_container_wait" git@github.com:jlhawn/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842354185800
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

jlhawn commented:
I've rebased after a conflict with `docs/api/version-history.md`.

jlhawn commented:
Since this didn't make the cut for 17.05 and API 1.29 I've had to add on an extra commit which bumps the version checks I added to use 1.30 instead and update `docs/api/version-history.md`.

tonistiigi commented on cli/command/container/wait.go:
```diff
@@ -28,6 +29,10 @@ func NewWaitCommand(dockerCli *command.DockerCli) *cobra.Command {
 	}
+
+	flags := cmd.Flags()
+	flags.StringVar(&opts.condition, "condition", "not-running", "Wait until a container state reaches the given condition, either 'not-running' (default), 'next-exit', or 'removed'")
```
Is there an use case for cli options? We don't necessarily need to expose the whole api in cli interface.

jlhawn replied:
I remember doing something in the past where I wanted to wait for a container on the CLI that was only "created" but I can't seem to remember why.

If we think it's just clutter then I can remove it, it would also remove the man-page changes and  some integration-cli tests that I added.

WeiZhang555 replied:
> **not-running**
> It blocks if the container is in a state which is considered to be running, i.e., any of running, paused, or restarting.

From your statement, I would expect that `docker restart` a running container won't make `wait --condition not-running` return, but actually it's not. After reading your codes, I see that `not-running` shares same `waitStop` channel with `next-exit`, the only difference is handling of initial state of container.

I think `not-running` is enough, not sure if there's any scenario for `next-exit`.

jlhawn replied:
`docker restart ...` and restart policies technically causes an atomic `stop` followed by a `start` (with a timeout before actually starting the process in the case of restart policies).

tonistiigi commented:
Design LGTM

ping @WeiZhang555 for review. Tentatively moving to code-review

cpuguy83 commented on cli/command/container/hijack.go:
```diff
@@ -28,6 +30,16 @@ func holdHijackedConnection(ctx context.Context, streams command.Streams, tty bo
+		escapeKeys, err := term.ToBytes(detachKeys)
+		if len(escapeKeys) == 0 || err != nil {
+			// Use defaults: ctrl-p, ctrl-q
+			escapeKeys = []byte{16, 17}
```
We need to capture this default somewhere.

cpuguy83 replied:
e.g., maybe package level var.

cpuguy83 commented on client/container_wait.go:
```diff
@@ -2,25 +2,85 @@ package client
-	resp, err := cli.post(ctx, "/containers/"+containerID+"/wait", nil, nil, nil)
-	if err != nil {
-		return -1, err
+// ContainerWait waits waits until the specified continer is in a certain state
```
waits waits

cpuguy83 commented on pkg/term/proxy.go:
```diff
@@ -0,0 +1,73 @@
+
+// NewProxy returns a new TTY proxy reader which wraps the given reader and
+// detects when the specified escape keys are read, in which case the
+func NewProxy(r io.Reader, escapeKeys []byte) io.Reader {
```
Probably not the best naming here.
`NewEscapeProxy`? `WithEscape`?

cpuguy83 commented on pkg/term/proxy.go:
```diff
@@ -0,0 +1,73 @@
+// ttyProxy is used only for attaches with a TTY. It is used to proxy
+// stdin keypresses from the underlying reader and look for the passed in
+// escape key sequence to signal a detach.
+type ttyProxy struct {
```
`escapeProxy`?

cpuguy83 commented on container/state.go:
```diff
@@ -335,6 +354,23 @@ func (s *State) SetDead() {
+}
+
 // Error returns current error for the state.
 func (s *State) Error() string {
```
🤦‍♂️

jlhawn replied:
I don't think this method is even used anymore, so I can just remove it.

cpuguy83 commented on container/state.go:
```diff
@@ -160,64 +165,78 @@ func IsValidStateString(s string) bool {
+//
+// WaitConditionRemoved is used to wait for the container to be removed.
+const (
+	WaitConditionNotRunning WaitCondition = iota
```
Maybe just need exited and removed?

jlhawn replied:
What is the definition of "exited" though?.

The original implementation of `wait` considered a container which was in the `"created"` state to be *exited* (`wait` for a `"created"` container returns with exit status 0 immediately) but we explicitly want to block when the container has not yet started, both in the case of `docker run ...` and `docker start -a ...` where the container may either be created or even already exited and you are about to manually start it again. The `next-exit` condition is used in these two cases. You can think about the `next-exit` condition as a state transition, or as the equivalent of waiting for a `die` event for the container.

cpuguy83 reviewed (changes requested):
Seems mostly alright.

jlhawn commented on api/server/router/container/container_routes.go:
```diff
@@ -279,13 +279,75 @@ func (s *containerRouter) postContainersUnpause(ctx context.Context, w http.Resp
+	// handler function signatures.
+	var closeNotify <-chan bool
+	if closeNotifier, ok := w.(http.CloseNotifier); ok {
+		closeNotify = closeNotifier.CloseNotify()
```
Thanks, this is good to know.

jlhawn commented:
@cpuguy83 I've rebased and adde 3 commits to address your comments, please take a look.

jlhawn commented:
@tonistiigi @WeiZhang555 I've also added a commit which removes the changes to the CLI, so there's no longer a `--condition` flag and it just keeps the old behavior (i.e., `--condition=not-running`). I also removed the integration tests which covered those CLI options.

jlhawn commented:
I used to be able to change labels on the `docker/docker` repo but now I can't. Can someone trigger a rebuild on `powerpc`, please?

cpuguy83 commented on client/container_wait.go:
```diff
@@ -2,25 +2,85 @@ package client
+
+// legacyContainerWait returns immediately and doesn't have an option to wait
+// until the container is removed.
+func (cli *Client) legacyContainerWait(ctx context.Context, containerID string) (<-chan container.ContainerWaitOKBody, <-chan error) {
```
It would be nice to consolidate this with the new function.
The new implementation looks more correct assuming we get a response immediately in both cases.

jlhawn replied:
We can't use the new client implementation because it makes the request and blocks on response headers *before* spawning a goroutine to read the body. There is also logic on the server side which will not return headers immediately if the API version is before `1.30`.

Even if the new server implementation did always return headers immediately, regardless of the API version, this would cause problems when talking to older servers which still block on the wait result before writing response headers.

To consolidate the two functions I'd have to add more version checks throughout it. Would you prefer that?

cpuguy83 commented on cli/command/container/utils.go:
```diff
@@ -13,12 +13,41 @@ import (
+		return legacyWaitExitOrRemoved(ctx, dockerCli, containerID, waitRemove)
+	}
+
+	condition := "next-exit"
```
Can we define these conditions in api/types?

jlhawn commented:
@cpuguy83 I've added a commit which adds `WaitCondition` constants to the `api/types/container` package. Please take a look.

jlhawn commented:
These build failures seem to be either Jenkins issues or flaky tests, can someone please trigger rebuilds for `experimental`, `janky`, and `windowsRS1`? 🙏

jlhawn commented:
Rebased after conflicts with `docs/api/version-history.md`.

tonistiigi commented on client/container_wait.go:
```diff
@@ -2,25 +2,83 @@ package client
+// wait request or in getting the response. This allows the caller to
+// sychronize ContainerWait with other calls, such as specifying a
+// "next-exit" condition before issuing a ContainerStart request.
+func (cli *Client) ContainerWait(ctx context.Context, containerID string, condition container.WaitCondition) (<-chan container.ContainerWaitOKBody, <-chan error) {
```
Why does this return channels? I don't see anything in here returning multiple items. As soon as anything is written to channels it closes down.

tiborvass replied:
Yeah why?

dnephin replied:
Response was here: https://github.com/moby/moby/pull/32237#issuecomment-301555524

tonistiigi commented on container/state.go:
```diff
@@ -160,64 +165,78 @@ func IsValidStateString(s string) bool {
+// its own, the result's Err() method wil be nil and its ExitCode() method will
+// return the conatiners exit code, otherwise, the results Err() method will
+// return an error indicating why the wait operation failed.
+func (s *State) Wait(ctx context.Context, condition WaitCondition) <-chan *StateStatus {
```
What needs this return channel? If we can avoid it there is no need for to rely on context canceling.

jlhawn replied:
So if it didn't return a channel then it would have to block? Wouldn't you still want to pass in a context to cancel that?

The reason this returns a channel (where `WaitStop` did not before) is because this call is usually wrapped by a call to `something.ContainerWait(ctx, containerID)` which returns the pair `(chan Status, error)`. If there is an error before the wait can begin then the error value will be non-nil. Once the wait begins though, the call no longer blocks and you get a channel back and the caller *knows that there is a backend goroutine blocking on the container to exit or be removed*. This allows the caller do do other things, synchronously, before pulling the value out of the channel, such as:

- create a container
- get channel to wait for that container's next exit
- start the container
- block on receiving container's exit status

If this did not return a channel than there would be race conditions in this use case:

- create container
- spawn goroutine to wait for container's next exit
- start container
- block on waiting for container's exit

In this scenario it's possible that you start the container and it exits before that previously spawned goroutine actually gets a chance to wait for the container and would then wait for the exit which would never happen again.

In the case of the API handler for `ContainerWait`, the returned `(chan Status, error)` values allow us to always return response headers immediately once the wait begins (or return `No Such Container` in the case of an error). That allows for the *"okay, the backend is now waiting, I'll send you the response once I have it"* flow to work through the API.

tonistiigi replied:
> So if it didn't return a channel then it would have to block? Wouldn't you still want to pass in a context to cancel that?

Canceling a context if a caller doesn't need a result anymore is what the `context.WithCancel` is meant for. Requiring in the implementation that a context passed into the function must be canceled in order to not leak resources is a different constraint.

This `Wait` has no way to return early and provide an early error. If it is more valuable than confusing that you can have an error with `200` status code that doesn't need to leak into lower-level functions. If I look at the client side code(and builder)
https://github.com/moby/moby/pull/32237/files#diff-316ca96dd5834654f8dc1537e5377bc3R28 then I don't see this property of HTTP api used. I don't any callers checking the status of the returned channel and error channel differently. So do we actually handle that race?

tonistiigi replied:
Anyway, if it is needed for avoiding a race we can keep it, just document all the functions that require context to be canceled.

jlhawn commented on container/state.go:
```diff
@@ -160,64 +165,78 @@ func IsValidStateString(s string) bool {
+	// actually stopped.
+	waitRemove := s.waitRemove
+
+	resultC := make(chan *StateStatus)
```
It doesn't need a buffer in this case since blocking on it doesn't prevent this function from returning. It's buffered in the other case (when the condition and container is `not-running`) because we can avoid needing to spawn a goroutine to put it in the channel.

tonistiigi replied:
Can't it still leak a goroutine?

jlhawn replied:
I guess another part of the contract with the caller would be that the caller *must* receive from the returned channel to avoid leaking any resources, since in this case the spawned goroutine does not exit until it is able to send on the channel. One could argue that the caller should always receive from the channel in which case it wouldn't matter whether the channel is buffered or not.

jlhawn commented on client/container_wait.go:
```diff
@@ -2,25 +2,83 @@ package client
+// wait request or in getting the response. This allows the caller to
+// sychronize ContainerWait with other calls, such as specifying a
+// "next-exit" condition before issuing a ContainerStart request.
+func (cli *Client) ContainerWait(ctx context.Context, containerID string, condition container.WaitCondition) (<-chan container.ContainerWaitOKBody, <-chan error) {
```
I guess this doesn't need to return an errors channel. The only reason why it does now is because there may be errors when decoding the response body. I could just log the error instead and set the response code to some non-zero value.

dnephin replied:
I have the same question.

Why does it return a `container.ContainerWaitOKBody` channel instead of just blocking?

jlhawn commented:
I've updated this PR again:
- Rebased to resolve yet another conflict with `docs/api/version-history.md`
- Added a new commit to address review comments from @tonistiigi

cpuguy83 commented:
```
15:49:24 The swagger spec at "api/swagger.yaml" is valid against swagger specification 2.0
15:49:26 The result of hack/generate-swagger-api.sh differs
15:49:26 
15:49:26  M api/types/container/container_wait.go
15:49:26 
15:49:26 Please update api/swagger.yaml with any api changes, then 
15:49:26 run `hack/generate-swagger-api.sh`.
```

cpuguy83 commented:
Ping @dnephin Do you know how we can add these new API objects to swagger so swagger-gen doesn't complain? Should we just move to a different file?

dnephin commented:
`ContainerWait` is a generated struct (it's in `hack/generate-swagger-api.sh`), so you should never edit `api/types/container/container_wait.go` directly. Edit swagger to match what you want and generate it using  `hack/generate-swagger-api.sh`.

Is that what you mean?

cpuguy83 commented:
I mean it makes sense, but he's added some enums which need to be defined somewhere, and preferably in one place, but not sure if there is any facility for that in the swagger spec/generator.

jlhawn commented:
> Edit swagger to match what you want ...

How would I do that? Edit the `api/swagger.yaml` file?

dnephin commented:
Yes, sorry I noticed afterward it was just an enum. Putting them into a different file should work.

jlhawn commented:
So, all of the `cli/command` files have disappeared. Do I have to open another PR now, against https://github.com/docker/cli ? I'm guessing that the api `client` package may soon disappear, too?

jlhawn commented:
I'm going to preserve my `cli/command` changes in this diff file and deal with that later. I've rebased and this PR is now strictly an `api` and backend change.


[moby_cli.diff.txt](https://github.com/moby/moby/files/980618/moby_cli.diff.txt)

cpuguy83 reviewed (approved):
LGTM if green.

jlhawn commented:
Tests *were* green, but I had to rebase to resolve still yet another conflict with `docs/api/version-history.md`.

jlhawn commented:
<img width="478" alt="screen shot 2017-05-11 at 10 51 02 am" src="https://cloud.githubusercontent.com/assets/1441929/25963883/c36deeac-3637-11e7-9f6e-9483fd3a0994.png">

💢💢💢💢💢💢💢💢💢💢💢💢

I'll keep rebasing to resolve the conflict, but I want to point out for posterity:

<img width="242" alt="screen shot 2017-05-11 at 10 51 42 am" src="https://cloud.githubusercontent.com/assets/1441929/25963925/f0bcbd16-3637-11e7-8847-6e1d5244b1b8.png">

WeiZhang555 commented:
The code implementation looks good to me. 

I have only one doubt about the `next-exit`, as I can see there's no function using the `next-exit` feature currently, then should we keep it in the codes ? Though it costs only little redundant codes. :cat2:

jlhawn commented:
> there's no function using the next-exit feature currently

There was, until the `cli` package was moved to https://github.com/docker/cli. See [this diff](https://github.com/moby/moby/files/980618/moby_cli.diff.txt) for the CLI changes for which I'll have to open another PR against that repository.

WeiZhang555 commented:
@jlhawn I see, LGTM :smile:

thaJeztah commented:
ping @tonistiigi @dnephin PTAL!

tonistiigi commented:
LGTM

dnephin commented on api/server/router/container/container_routes.go:
```diff
@@ -284,13 +284,53 @@ func (s *containerRouter) postContainersUnpause(ctx context.Context, w http.Resp
+	w.WriteHeader(http.StatusOK)
+	if flusher, ok := w.(http.Flusher); ok {
+		flusher.Flush()
+	}
```
Instead of duplicating the blocking on channel and returning lines I think you could put this block in an `if !legacyBehavior {...}`

jlhawn commented:
I'm currently rebasing and addressing @dnephin's comments. I want to expand on one of them since it has been asked already multiple times:

> Why does [the ContainerWait client method] return a container.ContainerWaitOKBody channel instead of just blocking?

This is done because it seemed like to quickest way to meet the new requirements we need for the Container Wait API which is to enable this flow:

1. Create a container in the case of `docker run ...` or have the container already exist in an exited state in the case of `docker start -a ...`.
2. Begin waiting for the container's *next* exit code (i.e., after it starts) or for it to be removed (in the case of `docker run --rm ...`.
3. Get *acknowledgement* from the daemon that it has begun waiting for that condition.
4. Start the container.
5. Continue to block on that wait condition.
6. Some time later, the container exits.
7. Receive the exit code.

- Point `2` is the reason why there is a `next-exit` wait condition.
- Point `3` is the reason why the API handler now returns response headers immediately.
- Being able to sync point `4` is the reason why the Client method returns a channel for the wait result - to be able to propagate the ack from point `3` and start the container without a race condition: If we didn't have a way of ensuring the order of `3` and `4` then it would be possible that the container is started and exits *before* the wait actually begins on the backend in which case `5` would block indefinitely.

I'm willing to change it to whatever we can all agree on. An alternative I have thought of is to go back to returning `(exitCode int, err error)` but to get the acknowledgement I mentioned we would need to have another argument to the method which is a channel which is maybe closed when the headers come back. That seems like an even more confusing interface to me though. Please make other suggestions if you have any ideas.

jlhawn commented:
I have no idea why the `windowsRS1` build has failed:
```
can't load package: package github.com/docker/cli/cmd/docker: cannot find package "github.com/docker/cli/cmd/docker" in any of:
 	C:\go\src\github.com\docker\cli\cmd\docker (from $GOROOT)
 	C:\Users\ContainerAdministrator\AppData\Local\Temp\e5237f98-d5ea-4be8-bfb3-c68f9d2c2d38\src\github.com\docker\cli\cmd\docker (from $GOPATH)
```

dnephin reviewed (approved):
LGTM

A bit light on tests, but I guess we can address that later.

Are all 12 commits still relevant? Do you think some of them could be squashed now that this is ready to merge?

jlhawn commented:
Rebased and squashed, it's now just 2 commits.

thaJeztah commented:
Thanks!

thaJeztah commented:
All 💚

jlhawn commented:
The CLI changes are now in a PR against the `docker/cli` repo here: https://github.com/docker/cli/pull/101
